kube-test/src/k3d.rs

```diff
@@ -0,0 +1,222 @@
+        );
+        assert_eq!(
+            self.kubeconfig
+                .clusters
+                .get(0)
+                .unwrap()
+                .name
+                .as_str()
+                .strip_prefix("k3d-")
+                .unwrap(),
+            self.name,
+            "kubeconfig only contains the temporary cluster"
+        );
+
+        let config = Config::from_custom_kubeconfig(self.kubeconfig.clone(), &Default::default())
+            .await
+            .expect("valid kubeconfig");
+        Client::try_from(config).expect("client")
+    }
+}
+
+impl Drop for TestEnv {
+    fn drop(&mut self) {
+        self.delete();
+    }
+}
+
+/// Builder for [`TestEnv`] to customize the environment.
+#[derive(Debug)]
+pub struct TestEnvBuilder {
+    /// The number of servers. Default: 1
+    servers: usize,
+    /// The number of agents. Default: 0
+    agents: usize,
+    /// Inject the Host IP as `host.k3d.internal` into the containers and CoreDNS.
+    /// Default: false
+    host_ip_injection: bool,
+    /// Create an image volume for importing images.
+    /// Default: false
+    create_image_volume: bool,
+    /// Create a `LoadBalancer` in front of the server nodes.
+    /// Default: false
+    create_load_balancer: bool,
+    /// Set `--verbose` flag. Default: false
+    verbose: bool,
+    /// Specify k3s version to use in format `v1.20.6`. Default: None.
+    version: Option<String>,
+}
+
+impl Default for TestEnvBuilder {
+    fn default() -> Self {
+        Self {
+            servers: 1,
+            agents: 0,
+            host_ip_injection: false,
+            create_image_volume: false,
+            create_load_balancer: false,
+            verbose: false,
+            version: None,
+        }
+    }
+}
+
+impl TestEnvBuilder {
+    /// Set the number of servers in the temporary cluster.
+    pub fn servers(&mut self, servers: usize) -> &mut Self {
+        self.servers = servers;
+        self
+    }
+
+    /// Set the number of agents in the temporary cluster.
+    pub fn agents(&mut self, agents: usize) -> &mut Self {
+        self.agents = agents;
+        self
+    }
+
+    /// Enable host ip injection.
+    pub fn inject_host_ip(&mut self) -> &mut Self {
+        self.host_ip_injection = true;
+        self
+    }
+
+    /// Create image volume.
+    pub fn with_image_volume(&mut self) -> &mut Self {
+        self.create_image_volume = true;
+        self
+    }
+
+    /// Create load balancer.
+    pub fn with_load_balancer(&mut self) -> &mut Self {
+        self.create_load_balancer = true;
+        self
+    }
+
+    /// Set `verbose` flag.
+    pub fn verbose(&mut self) -> &mut Self {
+        self.verbose = true;
+        self
+    }
+
+    /// Set the k3s version to use. The version should be in the form `v1.20.6`.
+    pub fn k3s_version<T: Into<String>>(&mut self, version: T) -> &mut Self {
+        self.version = Some(version.into());
+        self
+    }
+
+    /// Create the test environment.
```

Comment:
Maybe add links here/somewhere better to the variety of things that we are using arguments from:

- k3d cluster create: https://k3d.io/usage/commands/k3d_cluster_create/
- k3s server config: https://rancher.com/docs/k3s/latest/en/installation/install-options/server-config/